### PR TITLE
feat(drive): state-aware address funding fee estimation engine

### PR DESCRIPTION
**Pull request by llbartekll**
## Issue being fixed or feature implemented

Wallets funding platform addresses from an asset lock (0 address inputs, 1 remainder output) have no honest way to display the expected fee: the only client-side option is a static heuristic, and the server's own apply=false estimation prices the two data-dependent tree layers at `PotentiallyAtMaxElements` (32 levels), far above reality. This PR adds the missing primitive: a read-only, versioned `Drive::estimate_address_funding_fee` that prices the exact production operations against the current shape of the state trees.

**Stack (1/4):** engine → #4445 (query) → #4446 (client) → #4447 (FFI/Swift). This is the base PR.

## What was done?

- `Drive::estimate_address_funding_fee(recipient, outpoint, lock_credits, block_info, platform_version)` (versioned, `drive.methods.address_funds.estimate_funding_fee`):
  - builds a real `AddressFundingFromAssetLockTransitionActionV0` and runs it through the **production** high-level converter, then converts to low-level operations **statefully** (estimated layer info = `None`), so the insert-vs-replace branch for the recipient's balance and the element bytes come from committed state — byte-identical to apply=true execution. (In the apply=false path the stateless existence read returns `None` and the insert branch is taken unconditionally; the estimator must not inherit that.)
  - measures search-path levels for the recipient address (`[56,'c']`) and the asset lock outpoint (`[72]`) by generating single-key proofs against its own committed state and decoding them (new `util::proof_depth`, op semantics mirrored from `merk::proofs::tree::execute`); only the two data-dependent layer counts in the server's own `add_estimation_costs_*` maps are replaced with the measured levels — tree types, element sizes and all other layers stay the server model;
  - prices through the unchanged pipeline (`grove_batch_operations_costs` → `estimated_case_operations_for_batch` → `Drive::calculate_fee`). The result covers the batch only; validation operations and `user_fee_increase` are the caller's concern.
- v0 scope pinned fail-closed: a fresh, fully consumed lock only. An outpoint already present in `[72]` (fully or partially consumed) is rejected with the new `DriveError::AssetLockOutpointAlreadyPresent`, never silently priced.

## How Has This Been Tested?

Six new tests in `estimate_funding_fee/v0` (all green; full `cargo test -p drive --lib` = 3392 passed):
- read-only pin: grove root hash byte-identical before/after estimating (new address, existing address, rejected outpoint);
- stateful branch pin: new address builds `InsertOrReplace` with a zero nonce, existing address builds `Replace` with the summed balance;
- layer-map pin: differs from the server model only in the two measured counts;
- estimate vs real apply=true metered fee on the same state, empty and populated: empty 12,460,620 vs 12,551,520 (−0.7%), populated insert 13,175,300 vs 13,044,880 (+1.0%), replace 7,051,860 vs 6,802,640 (+3.7%), population 40 +1.0% — asserted within [85%, 115%] as regression headroom (documented as samples, not an upper-bound claim);
- present-outpoint rejection for both fully and partially consumed locks.

## Breaking Changes

None — additive (new method, new versioned field with value 0 in both existing instances, new error variant).

## Checklist:
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added or updated relevant unit/integration/functional/e2e tests
- [ ] I have added "!" to the title and described breaking changes in the corresponding section if my code contains any
- [ ] I have made corresponding changes to the documentation if needed

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

- **New Features**
  - Added state-aware fee estimation for funding an address from an asset lock.
  - Estimates account and asset-lock proof complexity for accurate fee results.
  - Reports recipient address existence and processing details.
  - Performs estimates without modifying committed state.
  - Supports current and legacy proof formats.
  - Supports versioned estimation across available platform configurations.

- **Bug Fixes**
  - Added clear errors when an asset-lock outpoint has already been used.
  - Prompts callers to retry if committed state changes during estimation.

### CODE REVIEW

File: packages/rs-drive/src/util/proof_depth.rs [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+        return Err(Error::Drive(DriveError::CorruptedDriveState(format!(
+            "local grovedb proof has {} trailing bytes",
+            proof_bytes.len() - consumed
+        ))));
+    }
+    let GroveDBProof::V1(v1) = proof else {
+        return Err(Error::Drive(DriveError::CorruptedDriveState(
+            "local grovedb proof is not a V1 envelope".to_string(),
+        )));
+    };
```
Comment (thepastaclaw) [inline packages/rs-drive/src/util/proof_depth.rs]: **🔴 Blocking: Support the V0 proof envelope used by protocol v11**

The decoder accepts only `GroveDBProof::V1`, but this PR enables `estimate_funding_fee: 0` in `DRIVE_ADDRESS_FUNDS_METHOD_VERSIONS_V1`, which protocol v11 selects through `DRIVE_VERSION_V6`. That Drive version pins `GROVE_V2`; its `prove_query_non_serialized` version is 0, and the pinned GroveDB implementation returns `GroveDBProof::V0` from that path. Therefore the first locally generated proof makes every protocol-v11 estimate fail with `CorruptedDriveState("local grovedb proof is not a V1 envelope")`. Address funding is active in v11, so traverse the V0 `MerkOnlyLayerProof` envelope as well as V1, and add a protocol-v11 regression test. All current estimator tests use `PlatformVersion::latest()`, so they do not exercise this supported version.

<sub>source: ['codex']</sub>

File: packages/rs-drive/src/drive/address_funds/estimate_funding_fee/v0/mod.rs [hunk 2]
```diff
@@ -0,0 +1,829 @@
...
+        // Measured search-path levels from locally generated proofs. GroveDB
+        // proving does not support transactions, so the whole estimate reads
+        // committed state.
+        let address_query = Drive::balance_for_clear_address_query(recipient);
+        let address_proof = self.grove_get_proved_path_query(
+            &address_query,
+            None,
+            &mut vec![],
+            &platform_version.drive,
+        )?;
```
Comment (thepastaclaw) [inline packages/rs-drive/src/drive/address_funds/estimate_funding_fee/v0/mod.rs:160]: **🟡 Suggestion: Take a coherent snapshot for all estimation reads**

The outpoint fetch, address proof, outpoint proof, and stateful low-level conversion are separate committed-state reads with no transaction or synchronization against block commits. A commit between the fetch and outpoint proof can legitimately make lines 171-175 report healthy state as `CorruptedDriveState`; returning `AssetLockOutpointAlreadyPresent` there would fix that error classification. More broadly, a recipient balance can change after its proof but before operation conversion, causing `address_exists` and the measured layer depth to describe one root while the generated insert/replace operation and fee describe another. Since the API promises state-aware pricing from committed state, serialize these reads with commits or compare the root before and after the complete operation and retry if it changed.

<sub>source: ['codex', 'coderabbit']</sub>

File: packages/rs-drive/src/util/proof_depth.rs [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+                let parent = pop_subtree(&mut stack)?;
+                let child = pop_subtree(&mut stack)?;
+                stack.push(attach_child(parent, child)?);
+            }
+            MerkProofOp::Child | MerkProofOp::ChildInverted => {
+                let child = pop_subtree(&mut stack)?;
+                let parent = pop_subtree(&mut stack)?;
+                stack.push(attach_child(parent, child)?);
+            }
+        }
```
Comment (thepastaclaw) [inline packages/rs-drive/src/util/proof_depth.rs:149]: **🟡 Suggestion: Add exact proof-depth tests for distinct search paths**

This custom interpreter independently reconstructs Merk's `Parent` and `Child` stack semantics, but the current tests never assert exact decoded depths. Populated absence cases always query key 200 after seeding keys 1 through N, so only an absence beyond the right edge is covered. Existing-address cases reach a present low key, but assert only presence and operation selection, while the ±15% fee bands can hide an off-by-one depth. Add direct tests using generated proofs for absent keys below the minimum, in an interior gap, and above the maximum, plus present keys on both sides and at different depths of a known small tree; assert the exact `SingleKeyProofLevels` values.

<sub>source: ['codex']</sub>

**Review by thepastaclaw — CHANGES_REQUESTED** [review]
## Preliminary review — Codex only

The estimator correctly reuses the production operation pipeline on the latest protocol version, but it cannot run under protocol v11 because its proof decoder rejects the V0 envelope produced by that version. The implementation also needs coherent committed-state reads and stronger exact-depth coverage for its custom Merk proof interpreter.
Source: reviewer backend model: gpt-5.6-sol; external CodeRabbit model: undisclosed; final verifier backend model: grok-4.5; orchestration-only: openclaw-agent/cliproxy/gpt-5.6-sol (not reviewer evidence).

Validated blockers were found in the Codex precheck. Opus is deferred until a fresh Codex revalidation clears the blocker gate.

### Review provenance
- Codex reviewers: `gpt-5.6-sol` — ffi-engineer (completed), `gpt-5.6-sol` — general (completed), `gpt-5.6-sol` — rust-quality (completed), `gpt-5.6-sol` — security-auditor (completed)
- Verifier: `gpt-5.6-sol` — verifier
- Sonnet: **not run (deferred by blocker gate)**

🔴 1 blocking | 🟡 2 suggestion(s)

<details>
<summary>🤖 Prompt for all review comments with AI agents</summary>

```
These findings are from an automated code review. Verify each finding against the current code and only fix it if needed.

In `packages/rs-drive/src/util/proof_depth.rs`:
- [BLOCKING] packages/rs-drive/src/util/proof_depth.rs:68-72: Support the V0 proof envelope used by protocol v11
  The decoder accepts only `GroveDBProof::V1`, but this PR enables `estimate_funding_fee: 0` in `DRIVE_ADDRESS_FUNDS_METHOD_VERSIONS_V1`, which protocol v11 selects through `DRIVE_VERSION_V6`. That Drive version pins `GROVE_V2`; its `prove_query_non_serialized` version is 0, and the pinned GroveDB implementation returns `GroveDBProof::V0` from that path. Therefore the first locally generated proof makes every protocol-v11 estimate fail with `CorruptedDriveState("local grovedb proof is not a V1 envelope")`. Address funding is active in v11, so traverse the V0 `MerkOnlyLayerProof` envelope as well as V1, and add a protocol-v11 regression test. All current estimator tests use `PlatformVersion::latest()`, so they do not exercise this supported version.
- [SUGGESTION] packages/rs-drive/src/util/proof_depth.rs:99-124: Add exact proof-depth tests for distinct search paths
  This custom interpreter independently reconstructs Merk's `Parent` and `Child` stack semantics, but the current tests never assert exact decoded depths. Populated absence cases always query key 200 after seeding keys 1 through N, so only an absence beyond the right edge is covered. Existing-address cases reach a present low key, but assert only presence and operation selection, while the ±15% fee bands can hide an off-by-one depth. Add direct tests using generated proofs for absent keys below the minimum, in an interior gap, and above the maximum, plus present keys on both sides and at different depths of a known small tree; assert the exact `SingleKeyProofLevels` values.

In `packages/rs-drive/src/drive/address_funds/estimate_funding_fee/v0/mod.rs`:
- [SUGGESTION] packages/rs-drive/src/drive/address_funds/estimate_funding_fee/v0/mod.rs:121-143: Take a coherent snapshot for all estimation reads
  The outpoint fetch, address proof, outpoint proof, and stateful low-level conversion are separate committed-state reads with no transaction or synchronization against block commits. A commit between the fetch and outpoint proof can legitimately make lines 171-175 report healthy state as `CorruptedDriveState`; returning `AssetLockOutpointAlreadyPresent` there would fix that error classification. More broadly, a recipient balance can change after its proof but before operation conversion, causing `address_exists` and the measured layer depth to describe one root while the generated insert/replace operation and fee describe another. Since the API promises state-aware pricing from committed state, serialize these reads with commits or compare the root before and after the complete operation and retry if it changed.
```
</details>

File: packages/rs-drive/src/drive/address_funds/estimate_funding_fee/v0/mod.rs [hunk 4]
```diff
@@ -0,0 +1,829 @@
...
+            &outpoint_proof,
+            &[asset_lock_storage_path()[0]],
+            asset_lock_outpoint.as_slice(),
+        )?;
+        if outpoint_levels.present {
+            return Err(Error::Drive(DriveError::CorruptedDriveState(
+                "asset lock outpoint reported absent by fetch but present in the local proof"
+                    .to_string(),
+            )));
+        }
```
[the remark this answers is not in the document]
Comment (thepastaclaw) [reply packages/rs-drive/src/drive/address_funds/estimate_funding_fee/v0/mod.rs:196]: The error classification issue is valid: the second read can observe an outpoint added by a normal concurrent commit, so this is not corrupted state. This is also one symptom of a broader snapshot problem: the address proof and subsequent stateful operation conversion can observe different roots, producing contradictory `address_exists`, operation-branch, and fee data.

File: packages/rs-drive/src/util/proof_depth.rs (hunk 1 above)
Comment (llbartekll) [reply packages/rs-drive/src/util/proof_depth.rs]: Fixed in 0d6fdc685e: the decoder now traverses the legacy `GroveDBProof::V0` (`MerkOnlyLayerProof`) envelope as well as V1, and a protocol-v11 regression test asserts the proof really is a V0 envelope (so the test keeps exercising the fixed path) and that the estimate still brackets the real apply=true fee under v11.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: packages/rs-drive/src/drive/address_funds/estimate_funding_fee/v0/mod.rs (hunk 2 above)
Comment (llbartekll) [reply packages/rs-drive/src/drive/address_funds/estimate_funding_fee/v0/mod.rs:160]: Fixed in 0d6fdc685e: the estimation's reads now count only when the grove root hash is byte-identical before and after all of them (up to 3 attempts); persistent instability returns a new retriable `CommittedStateChangedDuringOperation` error instead of mixed-root results, and the fetch/proof race is reclassified to `AssetLockOutpointAlreadyPresent`. The stacked #4445 maps the new error to a client-visible retry cue in the query handler.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: packages/rs-drive/src/util/proof_depth.rs (hunk 3 above)
Comment (llbartekll) [reply packages/rs-drive/src/util/proof_depth.rs:149]: Added in 0d6fdc685e: exact `SingleKeyProofLevels` assertions on generated proofs over known rotation-free AVL shapes (1/3/7 nodes — present keys on both sides at every depth; absences below the minimum, in every interior gap, and above the maximum), plus synthetic op-stream tests pinning the Parent/Child pop order, the absent-key +2 rule, and the underflow/leftover error paths.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: packages/rs-drive/src/drive/address_funds/estimate_funding_fee/v0/mod.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (llbartekll) [reply packages/rs-drive/src/drive/address_funds/estimate_funding_fee/v0/mod.rs:196]: Fixed in 0d6fdc685e: the fetch-absent/proof-present race now returns `AssetLockOutpointAlreadyPresent` — what the second read actually observed — instead of `CorruptedDriveState`, and the whole estimate is additionally guarded by a root-hash-stability retry loop.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: packages/rs-drive/src/error/drive.rs [hunk 5]
```diff
@@ -215,6 +215,17 @@ pub enum DriveError {
...
+    /// The asset lock outpoint is already present in the state (fully or
+    /// partially consumed), so an operation that requires a fresh outpoint
+    /// cannot proceed
+    #[error("asset lock outpoint already present in state: {0}")]
+    AssetLockOutpointAlreadyPresent(&'static str),
+
+    /// Committed state kept changing underneath a read-only operation that
+    /// needs several reads of one coherent snapshot; the caller should retry
+    #[error("committed state changed during a multi-read operation: {0}")]
+    CommittedStateChangedDuringOperation(&'static str),
```
[the remark this answers is not in the document]
Comment (llbartekll) [reply packages/rs-drive/src/error/drive.rs:227]: Not changed, per repo precedent: `DriveError` is routinely extended in ordinary releases — `NoCheckpointsAvailable`/`CheckpointNotFound` landed in #2917 (a plain `feat`, no breaking marker) and `PrefundedSpecializedBalanceNotEnough` in #1877 — with all workspace crates versioned in lockstep (4.2.0-dev). Flipping the enum to `#[non_exhaustive]` (or declaring a breaking release for two additive variants) would itself break downstream matches and is an API-policy decision for maintainers to take separately, not in this PR.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: packages/rs-drive/src/util/proof_depth.rs [hunk 6]
```diff
@@ -0,0 +1,468 @@
...
+    keyed: Vec<(u8, bool)>,
+}
+
+/// Decodes a locally generated GroveDB proof and returns the search-path
+/// structure for `key` in the merk layer at `path`.
+///
+/// `proof_bytes` must be the exact output of a single-key
+/// `grove_get_proved_path_query` over `path`/`key` on this node — the decode
+/// is canonical (trailing bytes rejected) and only `GroveDBProof::V1`
+/// envelopes with a plain merk terminal layer are supported.
```
Comment (thepastaclaw) [inline packages/rs-drive/src/util/proof_depth.rs]: **💬 Nitpick: Document both supported GroveDB proof envelopes**

The contract still says that only `GroveDBProof::V1` is supported, but the function now explicitly handles both the legacy V0 envelope and the current V1 envelope. This makes the protocol-v11 compatibility path appear accidental or unsupported even though it is covered by a regression test.

```suggestion
/// `grove_get_proved_path_query` over `path`/`key` on this node — the decode
/// is canonical (trailing bytes rejected); both legacy `GroveDBProof::V0` and
/// current `GroveDBProof::V1` envelopes with a plain merk terminal layer are
/// supported.
```

<sub>source: ['codex']</sub>

File: packages/rs-drive/src/drive/address_funds/estimate_funding_fee/v0/mod.rs [hunk 7]
```diff
@@ -0,0 +1,924 @@
...
+            _ => {
+                return Err(Error::Drive(DriveError::AssetLockOutpointAlreadyPresent(
+                    "address funding fee estimation requires a fresh (unspent) asset lock outpoint",
+                )));
+            }
+        }
+
+        // Measured search-path levels from locally generated proofs. GroveDB
+        // proving does not support transactions, so the whole estimate reads
+        // committed state.
```
Comment (thepastaclaw) [inline packages/rs-drive/src/drive/address_funds/estimate_funding_fee/v0/mod.rs]: **💬 Nitpick: Remove the incorrect claim that proof queries cannot use transactions**

`Drive::grove_get_proved_path_query` accepts a `TransactionArg` and forwards it to GroveDB, so the statement that proving does not support transactions is false. This estimator intentionally uses committed-state reads protected by the outer root-hash stability loop; the comment should describe that design instead.

```suggestion
// Measure search-path levels from locally generated proofs against committed
// state. The outer estimator accepts these reads only when the GroveDB root
// hash is identical before and after the complete state-dependent operation.
```

<sub>source: ['coderabbit']</sub>

**Review by thepastaclaw — COMMENTED** [review]
## Final validation — Codex/Sol only (Phase 2 disabled)

The estimator’s protocol-v11 proof handling, coherent-read guard, error classification, and exact proof-depth coverage are correct at the reviewed head. Two stale comments still contradict the implementation: one says only V1 proof envelopes are supported, and another incorrectly says GroveDB proving does not support transactions.
Source: reviewer backend model: gpt-5.6-sol; external CodeRabbit model: undisclosed; final verifier backend model: grok-4.5; orchestration-only: openclaw-agent/cliproxy/gpt-5.6-sol (not reviewer evidence).

Validated zero-blocker Codex/Sol precheck evidence was promoted to final because Phase 2 (Sonnet/Opus) is temporarily disabled. This is Codex/Sol-only final validation, not Codex + Sonnet/Opus coverage.

### Review provenance
- Codex reviewers: `gpt-5.6-sol` — general (completed), `gpt-5.6-sol` — rust-quality (completed)
- Verifier: `gpt-5.6-sol` — verifier
- Sonnet/Opus: **not run (Phase 2 disabled — temporary Codex/Sol-only final)**
- Secondary pass: disabled (`temporary_phase2_sonnet_disable`)

💬 2 nitpick(s)

<details>
<summary>🤖 Prompt for all review comments with AI agents</summary>

```
These findings are from an automated code review. Verify each finding against the current code and only fix it if needed.

In `packages/rs-drive/src/util/proof_depth.rs`:
- [NITPICK] packages/rs-drive/src/util/proof_depth.rs:43-46: Document both supported GroveDB proof envelopes
  The contract still says that only `GroveDBProof::V1` is supported, but the function now explicitly handles both the legacy V0 envelope and the current V1 envelope. This makes the protocol-v11 compatibility path appear accidental or unsupported even though it is covered by a regression test.

In `packages/rs-drive/src/drive/address_funds/estimate_funding_fee/v0/mod.rs`:
- [NITPICK] packages/rs-drive/src/drive/address_funds/estimate_funding_fee/v0/mod.rs:162-164: Remove the incorrect claim that proof queries cannot use transactions
  `Drive::grove_get_proved_path_query` accepts a `TransactionArg` and forwards it to GroveDB, so the statement that proving does not support transactions is false. This estimator intentionally uses committed-state reads protected by the outer root-hash stability loop; the comment should describe that design instead.
```
</details>

File: packages/rs-drive/src/drive/address_funds/estimate_funding_fee/v0/mod.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (thepastaclaw) [reply packages/rs-drive/src/drive/address_funds/estimate_funding_fee/v0/mod.rs:196]: The coherent-snapshot defect is resolved at the current head through the permitted root-before/root-after retry strategy rather than a read transaction. One part of this comment remains actionable: the source still incorrectly states that GroveDB proving does not support transactions, even though `grove_get_proved_path_query` accepts and forwards `TransactionArg`.

File: packages/rs-drive/src/error/drive.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (thepastaclaw) [reply packages/rs-drive/src/error/drive.rs:227]: The Rust exhaustiveness observation is technically true, but it does not establish an issue under this repository’s release policy. DriveError variants have repeatedly been added in ordinary non-breaking changes, including `NoCheckpointsAvailable` and `CheckpointNotFound` in 69032c948f and `PrefundedSpecializedBalanceNotEnough` in e7bfefa7a3, while workspace crates are versioned in lockstep.

File: packages/rs-drive/src/drive/address_funds/estimate_funding_fee/v0/mod.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (thepastaclaw) [reply packages/rs-drive/src/drive/address_funds/estimate_funding_fee/v0/mod.rs:196]: This is a mute/status reply and contains no technical finding to validate.

File: packages/rs-drive/src/util/proof_depth.rs (hunk 6 above)
Comment (llbartekll) [reply packages/rs-drive/src/util/proof_depth.rs]: Fixed in 8ff52d40b0 — the contract now documents both envelopes, with the applied suggestion extended by the why: protocol v11 selects GROVE_V2, whose prove path still emits V0.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: packages/rs-drive/src/drive/address_funds/estimate_funding_fee/v0/mod.rs (hunk 7 above)
Comment (llbartekll) [reply packages/rs-drive/src/drive/address_funds/estimate_funding_fee/v0/mod.rs]: Fixed in 8ff52d40b0 with the suggested wording (also mirrored in the dispatcher's doc comment): the comments now describe the actual design — committed-state reads accepted only when the GroveDB root hash is identical before and after the complete operation — instead of the transactions claim. Worth noting the repo does pin `prove_within_transaction_returns_not_supported` at runtime, but the design description is the right thing for this comment either way.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: packages/rs-drive/src/drive/address_funds/estimate_funding_fee/v0/mod.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (llbartekll) [reply packages/rs-drive/src/drive/address_funds/estimate_funding_fee/v0/mod.rs:196]: The stale transactions claim flagged here is corrected in 8ff52d40b0 (both in the parts function and the dispatcher doc), describing the committed-reads + root-hash-stability design instead.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: packages/rs-drive/src/drive/address_funds/estimate_funding_fee/v0/mod.rs [hunk 8]
```diff
@@ -0,0 +1,925 @@
...
+                .map_err(Error::from)?;
+            if root_before == root_after {
+                parts = Some(attempt);
+                break;
+            }
+        }
+        let Some((low_level_operations, layer_map, address_levels, outpoint_levels)) = parts else {
+            return Err(Error::Drive(DriveError::CommittedStateChangedDuringOperation(
+                "address funding fee estimation could not observe a stable committed state; retry",
+            )));
```
Comment (thepastaclaw) [inline packages/rs-drive/src/drive/address_funds/estimate_funding_fee/v0/mod.rs]: **🟡 Suggestion: Exercise the committed-state retry path**

This loop is the safeguard that prevents an estimate from combining reads made against different committed roots, but every current test uses a quiescent `Drive`, so only the first-attempt success branch executes. There is no deterministic coverage proving that an unstable attempt is discarded, that a later stable attempt is returned, or that three unstable attempts produce `DriveError::CommittedStateChangedDuringOperation`. Extract the root-stability loop into a small helper with controllable root samples or an injectable attempt closure, then pin all three behaviors so later changes cannot accidentally return stale attempt data or weaken the retry contract.

<sub>source: ['codex']</sub>

**Review by thepastaclaw — COMMENTED** [review]
## Final validation — Codex/Sol only (Phase 2 disabled)

The estimator now handles both supported GroveDB proof envelopes and protects its multi-read estimate with a committed-root stability check. No blocking defect remains, but the newly added retry and exhaustion behavior lacks deterministic test coverage, leaving one in-scope suggestion.
Source: reviewer backend model: gpt-5.6-sol; external CodeRabbit model: undisclosed; final verifier backend model: grok-4.5; orchestration-only: openclaw-agent/cliproxy/gpt-5.6-sol (not reviewer evidence).

Validated zero-blocker Codex/Sol precheck evidence was promoted to final because Phase 2 (Sonnet/Opus) is temporarily disabled. This is Codex/Sol-only final validation, not Codex + Sonnet/Opus coverage.

### Review provenance
- Codex reviewers: `gpt-5.6-sol` — general (completed), `gpt-5.6-sol` — rust-quality (completed)
- Verifier: `gpt-5.6-sol` — verifier
- Sonnet/Opus: **not run (Phase 2 disabled — temporary Codex/Sol-only final)**
- Secondary pass: disabled (`temporary_phase2_sonnet_disable`)

🟡 1 suggestion(s)

<details>
<summary>🤖 Prompt for all review comments with AI agents</summary>

```
These findings are from an automated code review. Verify each finding against the current code and only fix it if needed.

In `packages/rs-drive/src/drive/address_funds/estimate_funding_fee/v0/mod.rs`:
- [SUGGESTION] packages/rs-drive/src/drive/address_funds/estimate_funding_fee/v0/mod.rs:64-92: Exercise the committed-state retry path
  This loop is the safeguard that prevents an estimate from combining reads made against different committed roots, but every current test uses a quiescent `Drive`, so only the first-attempt success branch executes. There is no deterministic coverage proving that an unstable attempt is discarded, that a later stable attempt is returned, or that three unstable attempts produce `DriveError::CommittedStateChangedDuringOperation`. Extract the root-stability loop into a small helper with controllable root samples or an injectable attempt closure, then pin all three behaviors so later changes cannot accidentally return stale attempt data or weaken the retry contract.
```
</details>

File: packages/rs-drive/src/drive/address_funds/estimate_funding_fee/v0/mod.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (thepastaclaw) [reply packages/rs-drive/src/drive/address_funds/estimate_funding_fee/v0/mod.rs:196]: The coherent-read defect is fixed through the root-before/root-after retry strategy. A single read transaction is not viable here because GroveDB's proof path rejects transactional proving, as pinned by `prove_within_transaction_returns_not_supported`; the stale comment and outpoint error classification are also corrected at the exact head.

[the remark this answers is not in the document]
Comment (thepastaclaw) [reply packages/rs-drive/src/drive/address_funds/estimate_funding_fee/v0/mod.rs:196]: This is a mute/status reply and contains no technical finding to validate.

File: packages/rs-drive/src/error/drive.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (thepastaclaw) [reply packages/rs-drive/src/error/drive.rs:227]: This is a mute/status reply and contains no technical finding to validate.

File: packages/rs-drive/src/drive/address_funds/estimate_funding_fee/v0/mod.rs (hunk 8 above)
Comment (llbartekll) [reply packages/rs-drive/src/drive/address_funds/estimate_funding_fee/v0/mod.rs]: Added in 5d4e563442: the loop is extracted into `stable_committed_read(root_sample, attempt)` (behavior unchanged) and its contract is pinned deterministically with scripted root samples — a stable first attempt returns without a re-run, an unstable attempt's value is discarded in favor of a later stable one, three unstable attempts fail with the retriable `CommittedStateChangedDuringOperation` (exactly 3 attempts asserted), and an attempt error propagates immediately without a retry.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

**Review by thepastaclaw — APPROVED** [review]
## Final validation — Codex/Sol only (Phase 2 disabled)

The exact head resolves the prior retry-path coverage gap: `stable_committed_read` now has deterministic tests for first-attempt success, discarding an unstable result, exhaustion after exactly three attempts, and immediate error propagation; all four targeted tests passed locally. The CodeRabbit comments describe behavior already fixed at the exact head, an unsupported transaction-based alternative, repository-accepted `DriveError` evolution, or status-only replies, so no in-scope issue remains.
Source: reviewer backend model: gpt-5.6-sol; final verifier backend model: grok-4.5; orchestration-only: openclaw-agent/cliproxy/gpt-5.6-sol (not reviewer evidence).

Validated zero-blocker Codex/Sol precheck evidence was promoted to final because Phase 2 (Sonnet/Opus) is temporarily disabled. This is Codex/Sol-only final validation, not Codex + Sonnet/Opus coverage.

### Review provenance
- Codex reviewers: `gpt-5.6-sol` — general (completed), `gpt-5.6-sol` — security-auditor (completed), `gpt-5.6-sol` — rust-quality (completed)
- Verifier: `gpt-5.6-sol` — verifier
- Sonnet/Opus: **not run (Phase 2 disabled — temporary Codex/Sol-only final)**
- Secondary pass: disabled (`temporary_phase2_sonnet_disable`)

File: packages/rs-drive/src/drive/address_funds/estimate_funding_fee/v0/mod.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (thepastaclaw) [reply packages/rs-drive/src/drive/address_funds/estimate_funding_fee/v0/mod.rs:196]: A single read transaction is not viable for this path. Although the Drive wrapper accepts and forwards `TransactionArg`, GroveDB rejects proof generation inside a transaction, as pinned by `prove_within_transaction_returns_not_supported`. The exact head uses the supported root-stability retry strategy instead.

[the remark this answers is not in the document]
Comment (thepastaclaw) [reply packages/rs-drive/src/drive/address_funds/estimate_funding_fee/v0/mod.rs:196]: This is a mute/status response and contains no technical finding to validate.

File: packages/rs-drive/src/error/drive.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (thepastaclaw) [reply packages/rs-drive/src/error/drive.rs:227]: This reply correctly withdraws the preceding API-policy claim, but it is not itself an actionable finding.

[the remark this answers is not in the document]
Comment (thepastaclaw) [reply packages/rs-drive/src/error/drive.rs:227]: This is a mute/status response and contains no technical finding to validate.

**Review by QuantumExplorer — CHANGES_REQUESTED** [review]
Let's hold off on this till after we update various things in Platform to latest grovedb.
